<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../.resources/report.css" type="text/css"/><link rel="shortcut icon" href="../.resources/report.gif" type="image/gif"/><title>Codec.java</title><link rel="stylesheet" href="../.resources/prettify.css" type="text/css"/><script type="text/javascript" src="../.resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../.sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">org.jenetics</a> &gt; <a href="index.source.html" class="el_package">org.jenetics.engine</a> &gt; <span class="el_source">Codec.java</span></div><h1>Codec.java</h1><pre class="source lang-java linenums">/*
 * Java Genetic Algorithm Library (@__identifier__@).
 * Copyright (c) @__year__@ Franz Wilhelmstötter
 *
 * Licensed under the Apache License, Version 2.0 (the &quot;License&quot;);
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an &quot;AS IS&quot; BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Author:
 *    Franz Wilhelmstötter (franz.wilhelmstoetter@gmx.at)
 */
package org.jenetics.engine;

import static java.util.Objects.requireNonNull;

import java.util.function.BiFunction;
import java.util.function.Function;

import org.jenetics.Gene;
import org.jenetics.Genotype;
import org.jenetics.util.Factory;
import org.jenetics.util.ISeq;

/**
 * A problem {@code Codec} contains the information about how to encode a given
 * argument type into a {@code Genotype}. It also lets convert the encoded
 * {@code Genotype} back to the argument type. The engine creation and the
 * implementation of the fitness function can be heavily simplified by using
 * a {@code Codec} class. The example given in the {@link Engine} documentation
 * can be simplified as follows:
 *
 * &lt;pre&gt;{@code
 * public class RealFunction {
 *     // The conversion from the 'Genotype' to the argument type of the fitness
 *     // function is performed by the given 'Codec'. You can concentrate on the
 *     // implementation, because you are not bothered with the conversion code.
 *     private static double eval(final double x) {
 *         return cos(0.5 + sin(x)) * cos(x);
 *     }
 *
 *     public static void main(final String[] args) {
 *         final Engine&lt;DoubleGene, Double&gt; engine = Engine
 *              // Create an Engine.Builder with the &quot;pure&quot; fitness function
 *              // and the appropriate Codec.
 *             .build(RealFunction::eval, codecs.ofScalar(DoubleRange.of(0, 2*PI)))
 *             .build();
 *         ...
 *     }
 * }
 * }&lt;/pre&gt;
 *
 * The {@code Codec} needed for the above usage example, will look like this:
 * &lt;pre&gt;{@code
 * final DoubleRange domain = DoubleRange.of(0, 2*PI);
 * final Codec&lt;Double, DoubleGene&gt; codec = Codec.of(
 *     Genotype.of(DoubleChromosome.of(domain)),
 *     gt -&gt; gt.getChromosome().getGene().getAllele()
 * );
 * }&lt;/pre&gt;
 *
 * Calling the {@link Codec#of(Factory, Function)} method is the usual way for
 * creating new {@code Codec} instances.
 *
 * @see codecs
 * @see Engine
 * @see Engine.Builder
 *
 * @param &lt;T&gt; the argument type of a given problem
 * @param &lt;G&gt; the {@code Gene} type used for encoding the argument type {@code T}
 *
 * @author &lt;a href=&quot;mailto:franz.wilhelmstoetter@gmx.at&quot;&gt;Franz Wilhelmstötter&lt;/a&gt;
 * @version 3.3
 * @since 3.2
 */
public interface Codec&lt;T, G extends Gene&lt;?, G&gt;&gt; {

	/**
	 * Return the genotype factory for creating genotypes with the right
	 * encoding for the given problem. The genotype created with this factory
	 * must work together with the {@link #decoder()} function, which transforms
	 * the genotype into an object of the problem domain.
	 *
	 * &lt;pre&gt;{@code
	 * final Codec&lt;SomeObject, DoubleGene&gt; codec = ...
	 * final Genotype&lt;DoubleGene&gt; gt = codec.encoding().newInstance();
	 * final SomeObject arg = codec.decoder().apply(gt);
	 * }&lt;/pre&gt;
	 *
	 * @see #decoder()
	 *
	 * @return the genotype (factory) representation of the problem domain
	 */
	public Factory&lt;Genotype&lt;G&gt;&gt; encoding();

	/**
	 * Return the &lt;em&gt;decoder&lt;/em&gt; function which transforms the genotype back
	 * to the original problem domain representation.
	 *
	 * @see #encoding()
	 *
	 * @return genotype decoder
	 */
	public Function&lt;Genotype&lt;G&gt;, T&gt; decoder();


	/**
	 * Create a new {@code Codec} object with the given {@code encoding} and
	 * {@code decoder} function.
	 *
	 * @param encoding the genotype factory used for creating new
	 *        {@code Genotypes}.
	 * @param decoder decoder function, which converts a {@code Genotype} to a
	 *        value in the problem domain.
	 * @param &lt;G&gt; the {@code Gene} type
	 * @param &lt;T&gt; the fitness function argument type in the problem domain
	 * @return a new {@code Codec} object with the given parameters.
	 * @throws NullPointerException if one of the arguments is {@code null}.
	 */
	public static &lt;G extends Gene&lt;?, G&gt;, T&gt; Codec&lt;T, G&gt; of(
		final Factory&lt;Genotype&lt;G&gt;&gt; encoding,
		final Function&lt;Genotype&lt;G&gt;, T&gt; decoder
	) {
<span class="fc" id="L131">		requireNonNull(encoding);</span>
<span class="fc" id="L132">		requireNonNull(decoder);</span>

<span class="fc" id="L134">		return new Codec&lt;T, G&gt;() {</span>
			@Override
			public Factory&lt;Genotype&lt;G&gt;&gt; encoding() {
<span class="fc" id="L137">				return encoding;</span>
			}

			@Override
			public Function&lt;Genotype&lt;G&gt;, T&gt; decoder() {
<span class="fc" id="L142">				return decoder;</span>
			}
		};
	}


	/**
	 * Converts two given {@code Codec} instances into one. This lets you divide
	 * a problem into sub problems and combine them again.
	 * &lt;p&gt;
	 * The following example shows how to combine two codecs, which converts a
	 * {@code LongGene} to a {@code LocalDate}, to a codec which combines the
	 * two {@code LocalDate} object (this are the argument types of the
	 * component codecs) to a {@code Duration}.
	 *
	 * &lt;pre&gt;{@code
	 * final Codec&lt;LocalDate, LongGene&gt; dateCodec1 = Codec.of(
	 *     Genotype.of(LongChromosome.of(0, 10_000)),
	 *     gt -&gt; LocalDate.ofEpochDay(gt.getGene().longValue())
	 * );
	 *
	 * final Codec&lt;LocalDate, LongGene&gt; dateCodec2 = Codec.of(
	 *     Genotype.of(LongChromosome.of(1_000_000, 10_000_000)),
	 *     gt -&gt; LocalDate.ofEpochDay(gt.getGene().longValue())
	 * );
	 *
	 * final Codec&lt;Duration, LongGene&gt; durationCodec = Codec.of(
	 *     dateCodec1,
	 *     dateCodec2,
	 *     (d1, d2) -&gt; Duration.ofDays(d2.toEpochDay() - d1.toEpochDay())
	 * );
	 *
	 * final Engine&lt;LongGene, Long&gt; engine = Engine
	 *     .builder(Duration::toMillis, durationCodec)
	 *     .build();
	 *
	 * final Phenotype&lt;LongGene, Long&gt; pt = engine.stream()
	 *     .limit(100)
	 *     .collect(EvolutionResult.toBestPhenotype());
	 * System.out.println(pt);
	 *
	 * final Duration duration = durationCodec.decoder()
	 *     .apply(pt.getGenotype());
	 * System.out.println(duration);
	 * }&lt;/pre&gt;
	 *
	 * @since 3.3
	 *
	 * @param &lt;G&gt; the gene type
	 * @param &lt;A&gt; the argument type of the first codec
	 * @param &lt;B&gt; the argument type of the second codec
	 * @param &lt;T&gt; the argument type of the compound codec
	 * @param codec1 the first codec
	 * @param codec2 the second codec
	 * @param decoder the decoder which combines the two argument types from the
	 *        given given codecs, to the argument type of the resulting codec.
	 * @return a new codec which combines the given {@code codec1} and
	 *        {@code codec2}
	 * @throws NullPointerException if one of the arguments is {@code null}
	 */
	public static &lt;G extends Gene&lt;?, G&gt;, A, B, T&gt; Codec&lt;T, G&gt; of(
		final Codec&lt;A, G&gt; codec1,
		final Codec&lt;B, G&gt; codec2,
		final BiFunction&lt;A, B, T&gt; decoder
	) {
		@SuppressWarnings(&quot;unchecked&quot;)
<span class="fc" id="L208">		final Function&lt;Object[], T&gt; decoderAdapter = v -&gt; decoder</span>
			.apply((A)v[0], (B)v[1]);

<span class="fc" id="L211">		return of(</span>
<span class="fc" id="L212">			ISeq.of(codec1, codec2),</span>
			decoderAdapter
		);
	}

	/**
	 * Combines the given {@code codecs} into one codec. This lets you divide
	 * a problem into sub problems and combine them again.
	 * &lt;p&gt;
	 * The following example combines more than two sub-codecs into one.
	 * &lt;pre&gt;{@code
	 * final Codec&lt;LocalDate, LongGene&gt; dateCodec = Codec.of(
	 *     Genotype.of(LongChromosome.of(0, 10_000)),
	 *     gt -&gt; LocalDate.ofEpochDay(gt.getGene().longValue())
	 * );
	 *
	 * final Codec&lt;Duration, LongGene&gt; durationCodec = Codec.of(
	 *     ISeq.of(dateCodec, dateCodec, dateCodec),
	 *     dates -&gt; {
	 *         final LocalDate ld1 = (LocalDate)dates[0];
	 *         final LocalDate ld2 = (LocalDate)dates[1];
	 *         final LocalDate ld3 = (LocalDate)dates[2];
	 *
	 *         return Duration.ofDays(
	 *             ld1.toEpochDay() + ld2.toEpochDay() - ld3.toEpochDay()
	 *         );
	 *     }
	 * );
	 *
	 * final Engine&lt;LongGene, Long&gt; engine = Engine
	 *     .builder(Duration::toMillis, durationCodec)
	 *     .build();
	 *
	 * final Phenotype&lt;LongGene, Long&gt; pt = engine.stream()
	 *     .limit(100)
	 *     .collect(EvolutionResult.toBestPhenotype());
	 * System.out.println(pt);
	 *
	 * final Duration duration = durationCodec.decoder()
	 *     .apply(pt.getGenotype());
	 * System.out.println(duration);
	 * }&lt;/pre&gt;
	 *
	 * @since 3.3
	 *
	 * @param &lt;G&gt; the gene type
	 * @param &lt;T&gt; the argument type of the compound codec
	 * @param codecs the {@code Codec} sequence of the sub-problems
	 * @param decoder the decoder which combines the argument types from the
	 *        given given codecs, to the argument type of the resulting codec.
	 * @return a new codec which combines the given {@code codecs}
	 * @throws NullPointerException if one of the arguments is {@code null}
	 */
	public static &lt;G extends Gene&lt;?, G&gt;, T&gt; Codec&lt;T, G&gt; of(
		final ISeq&lt;? extends Codec&lt;?, G&gt;&gt; codecs,
		final Function&lt;? super Object[], ? extends T&gt; decoder
	) {
<span class="fc" id="L269">		return new CompositeCodec&lt;&gt;(codecs, decoder);</span>
	}

}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.1.201405082137</span></div></body></html>