<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../.resources/report.css" type="text/css"/><link rel="shortcut icon" href="../.resources/report.gif" type="image/gif"/><title>CharacterChromosome.java</title><link rel="stylesheet" href="../.resources/prettify.css" type="text/css"/><script type="text/javascript" src="../.resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../.sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">org.jenetics</a> &gt; <a href="index.source.html" class="el_package">org.jenetics</a> &gt; <span class="el_source">CharacterChromosome.java</span></div><h1>CharacterChromosome.java</h1><pre class="source lang-java linenums">/*
 * Java Genetic Algorithm Library (@__identifier__@).
 * Copyright (c) @__year__@ Franz Wilhelmstötter
 *
 * Licensed under the Apache License, Version 2.0 (the &quot;License&quot;);
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an &quot;AS IS&quot; BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Author:
 *    Franz Wilhelmstötter (franz.wilhelmstoetter@gmx.at)
 */
package org.jenetics;

import static org.jenetics.CharacterGene.DEFAULT_CHARACTERS;
import static org.jenetics.internal.util.Equality.eq;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.util.function.Supplier;
import java.util.stream.Collectors;

import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorType;
import javax.xml.bind.annotation.XmlAttribute;
import javax.xml.bind.annotation.XmlElement;
import javax.xml.bind.annotation.XmlRootElement;
import javax.xml.bind.annotation.XmlType;
import javax.xml.bind.annotation.adapters.XmlAdapter;
import javax.xml.bind.annotation.adapters.XmlJavaTypeAdapter;

import org.jenetics.internal.util.Equality;
import org.jenetics.internal.util.Hash;
import org.jenetics.internal.util.IntRef;

import org.jenetics.util.CharSeq;
import org.jenetics.util.ISeq;
import org.jenetics.util.MSeq;

/**
 * CharacterChromosome which represents character sequences.
 *
 * @author &lt;a href=&quot;mailto:franz.wilhelmstoetter@gmx.at&quot;&gt;Franz Wilhelmstötter&lt;/a&gt;
 * @since 1.0
 * @version 3.0
 */
@XmlJavaTypeAdapter(CharacterChromosome.Model.Adapter.class)
public class CharacterChromosome
	extends
		AbstractChromosome&lt;CharacterGene&gt;
	implements
		CharSequence,
		Serializable
{
	private static final long serialVersionUID = 2L;

	private transient CharSeq _validCharacters;

	/**
	 * Create a new chromosome from the given {@code genes} array. The genes
	 * array is copied, so changes to the given genes array doesn't effect the
	 * genes of this chromosome.
	 *
	 * @param genes the genes that form the chromosome.
	 * @throws NullPointerException if the given gene array is {@code null}.
	 * @throws IllegalArgumentException if the length of the gene array is
	 *         smaller than one.
	 */
	protected CharacterChromosome(final ISeq&lt;CharacterGene&gt; genes) {
<span class="fc" id="L79">		super(genes);</span>
<span class="fc" id="L80">		_validCharacters = genes.get(0).getValidCharacters();</span>
<span class="fc" id="L81">	}</span>

	/**
	 * Create a new chromosome with the {@code validCharacters} char set as
	 * valid characters.
	 *
	 * @param validCharacters the valid characters for this chromosome.
	 * @param length the length of the new chromosome.
	 * @throws NullPointerException if the {@code validCharacters} is
	 *         {@code null}.
	 * @throws IllegalArgumentException if the {@code length} is smaller than
	 *         one.
	 */
	public CharacterChromosome(final CharSeq validCharacters, final int length) {
<span class="fc" id="L95">		this(CharacterGene.seq(validCharacters, length));</span>
<span class="fc" id="L96">		_valid = true;</span>
<span class="fc" id="L97">	}</span>

	@Override
	public char charAt(final int index) {
<span class="fc" id="L101">		return getGene(index).charValue();</span>
	}

	@Override
	public CharacterChromosome subSequence(final int start, final int end) {
<span class="nc" id="L106">		return new CharacterChromosome(_genes.subSeq(start, end));</span>
	}

	/**
	 * @throws NullPointerException if the given gene array is {@code null}.
	 */
	@Override
	public CharacterChromosome newInstance(final ISeq&lt;CharacterGene&gt; genes) {
<span class="fc" id="L114">		return new CharacterChromosome(genes);</span>
	}

	/**
	 * Create a new, &lt;em&gt;random&lt;/em&gt; chromosome.
	 */
	@Override
	public CharacterChromosome newInstance() {
<span class="fc" id="L122">		return new CharacterChromosome(_validCharacters, length());</span>
	}

	@Override
	public int hashCode() {
<span class="fc" id="L127">		return Hash.of(getClass())</span>
<span class="fc" id="L128">				.and(super.hashCode())</span>
<span class="fc" id="L129">				.and(_validCharacters).value();</span>
	}

	@Override
	public boolean equals(final Object obj) {
<span class="fc" id="L134">		return Equality.of(this, obj).test(cc -&gt;</span>
			super.equals(obj) &amp;&amp;
			eq(_validCharacters, cc._validCharacters)
		);
	}

	@Override
	public String toString() {
<span class="fc" id="L142">		return toSeq().stream()</span>
<span class="fc" id="L143">			.map(Object::toString)</span>
<span class="fc" id="L144">			.collect(Collectors.joining());</span>
	}

	/**
	 * Returns an char array containing all of the elements in this chromosome
	 * in proper sequence.  If the chromosome fits in the specified array, it is
	 * returned therein. Otherwise, a new array is allocated with the length of
	 * this chromosome.
	 *
	 * @since 3.0
	 *
	 * @param array the array into which the elements of this chromosomes are to
	 *        be stored, if it is big enough; otherwise, a new array is
	 *        allocated for this purpose.
	 * @return an array containing the elements of this chromosome
	 * @throws NullPointerException if the given {@code array} is {@code null}
	 */
	public char[] toArray(final char[] array) {
<span class="nc bnc" id="L162" title="All 2 branches missed.">		final char[] a = array.length &gt;= length() ?</span>
<span class="nc" id="L163">			array : new char[length()];</span>

<span class="nc bnc" id="L165" title="All 2 branches missed.">		for (int i = length(); --i &gt;= 0;) {</span>
<span class="nc" id="L166">			a[i] = charAt(i);</span>
		}

<span class="nc" id="L169">		return a;</span>
	}

	/**
	 * Returns an char array containing all of the elements in this chromosome
	 * in proper sequence.
	 *
	 * @since 3.0
	 *
	 * @return an array containing the elements of this chromosome
	 */
	public char[] toArray() {
<span class="nc" id="L181">		return toArray(new char[length()]);</span>
	}


	/**
	 * Create a new chromosome with the {@link CharacterGene#DEFAULT_CHARACTERS}
	 * char set as valid characters.
	 *
	 * @param length the {@code length} of the new chromosome.
	 * @return a new {@code CharacterChromosome} with the given parameter
	 * @throws IllegalArgumentException if the {@code length} is smaller than
	 *         one.
	 */
	public static CharacterChromosome of(final int length) {
<span class="fc" id="L195">		return new CharacterChromosome(</span>
<span class="fc" id="L196">			CharacterGene.seq(DEFAULT_CHARACTERS, length)</span>
		);
	}

	/**
	 * Create a new chromosome from the given genes (given as string).
	 *
	 * @param alleles the character genes.
	 * @param validChars the valid characters.
	 * @return a new {@code CharacterChromosome} with the given parameter
	 * @throws IllegalArgumentException if the genes string is empty.
	 */
	public static CharacterChromosome of(
		final String alleles,
		final CharSeq validChars
	) {
<span class="fc" id="L212">		final IntRef index = new IntRef();</span>
<span class="fc" id="L213">		final Supplier&lt;CharacterGene&gt; geneFactory = () -&gt; CharacterGene.of(</span>
			alleles.charAt(index.value++), validChars
		);

<span class="fc" id="L217">		final ISeq&lt;CharacterGene&gt; genes =</span>
<span class="fc" id="L218">			MSeq.&lt;CharacterGene&gt;ofLength(alleles.length())</span>
<span class="fc" id="L219">				.fill(geneFactory)</span>
<span class="fc" id="L220">				.toISeq();</span>

<span class="fc" id="L222">		return new CharacterChromosome(genes);</span>
	}

	/**
	 * Create a new chromosome from the given genes (given as string).
	 *
	 * @param alleles the character genes.
	 * @return a new {@code CharacterChromosome} with the given parameter
	 * @throws IllegalArgumentException if the genes string is empty.
	 */
	public static CharacterChromosome of(final String alleles) {
<span class="nc" id="L233">		return of(alleles, DEFAULT_CHARACTERS);</span>
	}


	/* *************************************************************************
	 *  Java object serialization
	 * ************************************************************************/

	private void writeObject(final ObjectOutputStream out)
		throws IOException
	{
<span class="fc" id="L244">		out.defaultWriteObject();</span>

<span class="fc" id="L246">		out.writeInt(length());</span>
<span class="fc" id="L247">		out.writeObject(_validCharacters);</span>

<span class="fc bfc" id="L249" title="All 2 branches covered.">		for (CharacterGene gene : _genes) {</span>
<span class="fc" id="L250">			out.writeChar(gene.getAllele().charValue());</span>
<span class="fc" id="L251">		}</span>
<span class="fc" id="L252">	}</span>

	private void readObject(final ObjectInputStream in)
		throws IOException, ClassNotFoundException
	{
<span class="fc" id="L257">		in.defaultReadObject();</span>

<span class="fc" id="L259">		final int length = in.readInt();</span>
<span class="fc" id="L260">		_validCharacters = (CharSeq)in.readObject();</span>

<span class="fc" id="L262">		final MSeq&lt;CharacterGene&gt; genes = MSeq.ofLength(length);</span>
<span class="fc bfc" id="L263" title="All 2 branches covered.">		for (int i = 0; i &lt; length; ++i) {</span>
<span class="fc" id="L264">			final CharacterGene gene = CharacterGene.of(</span>
<span class="fc" id="L265">				in.readChar(),</span>
				_validCharacters
			);
<span class="fc" id="L268">			genes.set(i, gene);</span>
		}

<span class="fc" id="L271">		_genes = genes.toISeq();</span>
<span class="fc" id="L272">	}</span>

	/* *************************************************************************
	 *  JAXB object serialization
	 * ************************************************************************/

	@XmlRootElement(name = &quot;character-chromosome&quot;)
	@XmlType(name = &quot;org.jenetics.CharacterChromosome&quot;)
	@XmlAccessorType(XmlAccessType.FIELD)
<span class="fc" id="L281">	final static class Model {</span>

		@XmlAttribute(name = &quot;length&quot;, required = true)
		public int length;

		@XmlElement(name = &quot;valid-alleles&quot;, required = true, nillable = false)
		public String validCharacters;

		@XmlElement(name = &quot;alleles&quot;, required = true, nillable = false)
		public String genes;

<span class="fc" id="L292">		public final static class Adapter</span>
			extends XmlAdapter&lt;Model, CharacterChromosome&gt;
		{
			@Override
			public Model marshal(final CharacterChromosome value) {
<span class="nc" id="L297">				final Model m = new Model();</span>
<span class="nc" id="L298">				m.length = value.length();</span>
<span class="nc" id="L299">				m.validCharacters = value._validCharacters.toString();</span>
<span class="nc" id="L300">				m.genes = value.toString();</span>
<span class="nc" id="L301">				return m;</span>
			}

			@Override
			public CharacterChromosome unmarshal(final Model m) {
<span class="fc" id="L306">				return CharacterChromosome.of(</span>
					m.genes,
					new CharSeq(m.validCharacters)
				);
			}
		}
	}
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.1.201405082137</span></div></body></html>