<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../.resources/report.css" type="text/css"/><link rel="shortcut icon" href="../.resources/report.gif" type="image/gif"/><title>EnumGene.java</title><link rel="stylesheet" href="../.resources/prettify.css" type="text/css"/><script type="text/javascript" src="../.resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../.sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">org.jenetics</a> &gt; <a href="index.source.html" class="el_package">org.jenetics</a> &gt; <span class="el_source">EnumGene.java</span></div><h1>EnumGene.java</h1><pre class="source lang-java linenums">/*
 * Java Genetic Algorithm Library (@__identifier__@).
 * Copyright (c) @__year__@ Franz Wilhelmstötter
 *
 * Licensed under the Apache License, Version 2.0 (the &quot;License&quot;);
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an &quot;AS IS&quot; BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Author:
 *    Franz Wilhelmstötter (franz.wilhelmstoetter@gmx.at)
 */
package org.jenetics;

import static java.lang.String.format;
import static org.jenetics.internal.util.Equality.eq;

import java.io.Serializable;
import java.util.List;
import java.util.Objects;
import java.util.function.Function;

import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorType;
import javax.xml.bind.annotation.XmlAttribute;
import javax.xml.bind.annotation.XmlElement;
import javax.xml.bind.annotation.XmlElementWrapper;
import javax.xml.bind.annotation.XmlRootElement;
import javax.xml.bind.annotation.XmlType;
import javax.xml.bind.annotation.adapters.XmlAdapter;
import javax.xml.bind.annotation.adapters.XmlJavaTypeAdapter;

import org.jenetics.internal.util.Hash;
import org.jenetics.internal.util.jaxb;
import org.jenetics.internal.util.model.IndexedObject;
import org.jenetics.internal.util.reflect;

import org.jenetics.util.ISeq;
import org.jenetics.util.RandomRegistry;

/**
 * &lt;p&gt;
 * Gene which holds enumerable (countable) genes. Will be used for combinatorial
 * problems in combination with the {@link PermutationChromosome}.
 * &lt;/p&gt;
 * The following code shows how to create a combinatorial genotype factory which
 * can be used when creating an {@link org.jenetics.engine.Engine} instance.
 * &lt;pre&gt;{@code
 * final ISeq&lt;Integer&gt; alleles = ISeq.of(1, 2, 3, 4, 5, 6, 7, 8);
 * final Factory&lt;Genotype&lt;EnumGene&lt;Integer&gt;&gt;&gt; gtf = Genotype.of(
 *     PermutationChromosome.of(alleles)
 * );
 * }&lt;/pre&gt;
 *
 * The following code shows the assurances of the {@code EnumGene}.
 * &lt;pre&gt;{@code
 * final ISeq&lt;Integer&gt; alleles = ISeq.of(1, 2, 3, 4, 5, 6, 7, 8);
 * final EnumGene&lt;Integer&gt; gene = new EnumGene&lt;&gt;(5, alleles);
 *
 * assert(gene.getAlleleIndex() == 5);
 * assert(gene.getAllele() == gene.getValidAlleles().get(5));
 * assert(gene.getValidAlleles() == alleles);
 * }&lt;/pre&gt;
 *
 * @see PermutationChromosome
 *
 * @author &lt;a href=&quot;mailto:franz.wilhelmstoetter@gmx.at&quot;&gt;Franz Wilhelmstötter&lt;/a&gt;
 * @since 1.0
 * @version 2.0
 */
@XmlJavaTypeAdapter(EnumGene.Model.Adapter.class)
public final class EnumGene&lt;A&gt;
	implements
		Gene&lt;A, EnumGene&lt;A&gt;&gt;,
		Comparable&lt;EnumGene&lt;A&gt;&gt;,
		Serializable
{

	private static final long serialVersionUID = 2L;

	private final ISeq&lt;A&gt; _validAlleles;
	private final int _alleleIndex;

	/**
	 * Create a new enum gene from the given valid genes and the chosen allele
	 * index.
	 * @param alleleIndex the index of the allele for this gene.
	 * @param validAlleles the sequence of valid alleles.
	 * @throws IllegalArgumentException if the give valid alleles sequence is
	 *         empty
	 * @throws NullPointerException if the valid alleles seq is {@code null}.
	 */
<span class="fc" id="L100">	EnumGene(final int alleleIndex, final ISeq&lt;? extends A&gt; validAlleles) {</span>
<span class="pc bpc" id="L101" title="1 of 2 branches missed.">		if (validAlleles.length() == 0) {</span>
<span class="nc" id="L102">			throw new IllegalArgumentException(</span>
				&quot;Array of valid alleles must be greater than zero.&quot;
			);
		}

<span class="fc bfc" id="L107" title="All 4 branches covered.">		if (alleleIndex &lt; 0 || alleleIndex &gt;= validAlleles.length()) {</span>
<span class="fc" id="L108">			throw new IndexOutOfBoundsException(format(</span>
<span class="fc" id="L109">				&quot;Allele index is not in range [0, %d).&quot;, alleleIndex</span>
			));
		}

<span class="fc" id="L113">		_validAlleles = reflect.cast(validAlleles);</span>
<span class="fc" id="L114">		_alleleIndex = alleleIndex;</span>
<span class="fc" id="L115">	}</span>

	/**
	 * Return sequence of the valid alleles where this gene is a part of.
	 *
	 * @return the sequence of the valid alleles.
	 */
	public ISeq&lt;A&gt; getValidAlleles() {
<span class="fc" id="L123">		return _validAlleles;</span>
	}

	/**
	 * Return the index of the allele this gene is representing.
	 *
	 * @return the index of the allele this gene is representing.
	 */
	public int getAlleleIndex() {
<span class="fc" id="L132">		return _alleleIndex;</span>
	}

	@Override
	public A getAllele() {
<span class="fc" id="L137">		return _validAlleles.get(_alleleIndex);</span>
	}

	@Override
	public boolean isValid() {
<span class="pc bpc" id="L142" title="2 of 4 branches missed.">		return _alleleIndex &gt;= 0 &amp;&amp; _alleleIndex &lt; _validAlleles.length();</span>
	}

	@Override
	public EnumGene&lt;A&gt; newInstance() {
<span class="fc" id="L147">		return new EnumGene&lt;&gt;(</span>
<span class="fc" id="L148">			RandomRegistry.getRandom().nextInt(_validAlleles.length()),</span>
			_validAlleles
		);
	}

	/**
	 * Create a new gene from the given {@code value} and the gene context.
	 *
	 * @since 1.6
	 * @param value the value of the new gene.
	 * @return a new gene with the given value.
	 */
	public EnumGene&lt;A&gt; newInstance(final A value) {
<span class="nc" id="L161">		return new EnumGene&lt;&gt;(</span>
<span class="nc" id="L162">			_validAlleles.indexOf(value),</span>
			_validAlleles
		);
	}

	@Override
	public int compareTo(final EnumGene&lt;A&gt; gene) {
<span class="nc" id="L169">		int result = 0;</span>
<span class="nc bnc" id="L170" title="All 2 branches missed.">		if (_alleleIndex &gt; gene._alleleIndex) {</span>
<span class="nc" id="L171">			result = 1;</span>
<span class="nc bnc" id="L172" title="All 2 branches missed.">		} else if (_alleleIndex &lt; gene._alleleIndex) {</span>
<span class="nc" id="L173">			result = -1;</span>
		}

<span class="nc" id="L176">		return result;</span>
	}

	@Override
	public int hashCode() {
<span class="fc" id="L181">		return Hash.of(EnumGene.class)</span>
<span class="fc" id="L182">				.and(_alleleIndex)</span>
<span class="fc" id="L183">				.and(_validAlleles).value();</span>
	}

	@Override
	public boolean equals(final Object obj) {
<span class="fc bfc" id="L188" title="All 2 branches covered.">		return obj instanceof EnumGene &amp;&amp;</span>
<span class="fc bfc" id="L189" title="All 2 branches covered.">			eq(((EnumGene)obj)._alleleIndex, _alleleIndex) &amp;&amp;</span>
<span class="pc bpc" id="L190" title="1 of 2 branches missed.">			eq(((EnumGene)obj)._validAlleles, _validAlleles);</span>
	}

	@Override
	public String toString() {
<span class="fc" id="L195">		return Objects.toString(getAllele());</span>
	}

	/* *************************************************************************
	 *  Static object creation methods
	 * ************************************************************************/

	/**
	 * Return a new enum gene with an allele randomly chosen from the given
	 * valid alleles.
	 *
	 * @param &lt;A&gt; the allele type
	 * @param validAlleles the sequence of valid alleles.
	 * @return a new {@code EnumGene} with the given parameter
	 * @throws java.lang.IllegalArgumentException if the give valid alleles
	 *         sequence is empty
	 * @throws NullPointerException if the valid alleles seq is {@code null}.
	 */
	public static &lt;A&gt; EnumGene&lt;A&gt; of(final ISeq&lt;? extends A&gt; validAlleles) {
<span class="fc" id="L214">		return new EnumGene&lt;&gt;(</span>
<span class="fc" id="L215">			RandomRegistry.getRandom().nextInt(validAlleles.length()),</span>
			validAlleles
		);
	}

	/**
	 * Create a new enum gene from the given valid genes and the chosen allele
	 * index.
	 *
	 * @param &lt;A&gt; the allele type
	 * @param alleleIndex the index of the allele for this gene.
	 * @param validAlleles the array of valid alleles.
	 * @return a new {@code EnumGene} with the given parameter
	 * @throws java.lang.IllegalArgumentException if the give valid alleles
	 *         array is empty of the allele index is out of range.
	 */
	@SafeVarargs
	public static &lt;A&gt; EnumGene&lt;A&gt; of(
		final int alleleIndex,
		final A... validAlleles
	) {
<span class="nc" id="L236">		return new EnumGene&lt;&gt;(alleleIndex, ISeq.of(validAlleles));</span>
	}

	/**
	 * Return a new enum gene with an allele randomly chosen from the given
	 * valid alleles.
	 *
	 * @param &lt;A&gt; the allele type
	 * @param validAlleles the array of valid alleles.
	 * @return a new {@code EnumGene} with the given parameter
	 * @throws java.lang.IllegalArgumentException if the give valid alleles
	 *         array is empty
	 */
	@SafeVarargs
	public static &lt;A&gt; EnumGene&lt;A&gt; of(final A... validAlleles) {
<span class="nc" id="L251">		return EnumGene.of(ISeq.of(validAlleles));</span>
	}

	/* *************************************************************************
	 *  JAXB object serialization
	 * ************************************************************************/

	@XmlRootElement(name = &quot;enum-gene&quot;)
	@XmlType(name = &quot;org.jenetics.EnumGene&quot;)
	@XmlAccessorType(XmlAccessType.FIELD)
	@SuppressWarnings({&quot;unchecked&quot;, &quot;rawtypes&quot;})
<span class="fc" id="L262">	final static class Model {</span>

		@XmlAttribute(name = &quot;length&quot;, required = true)
		public int length;

		@XmlElementWrapper(name = &quot;valid-alleles&quot;, required = true, nillable = false)
		@XmlElement(name = &quot;allele&quot;, required = true, nillable = false)
		public List alleles;

<span class="fc" id="L271">		@XmlElement(name = &quot;allele&quot;, required = true, nillable = false)</span>
		public IndexedObject allele = new IndexedObject();

<span class="fc" id="L274">		public static final class Adapter</span>
			extends XmlAdapter&lt;Model, EnumGene&gt;
		{
			@Override
			public Model marshal(final EnumGene gene) {
<span class="nc" id="L279">				final Function marshaller = jaxb.Marshaller(gene.getAllele());</span>
<span class="nc" id="L280">				final Model m = new Model();</span>
<span class="nc" id="L281">				m.length = gene.getValidAlleles().length();</span>
<span class="nc" id="L282">				m.allele.index = gene.getAlleleIndex();</span>
<span class="nc" id="L283">				m.allele.value = marshaller.apply(gene.getAllele());</span>
<span class="nc" id="L284">				m.alleles = gene.getValidAlleles()</span>
<span class="nc" id="L285">					.map(marshaller)</span>
<span class="nc" id="L286">					.asList();</span>

<span class="nc" id="L288">				return m;</span>
			}

			@Override
			public EnumGene unmarshal(final Model m) {
<span class="fc" id="L293">				return new EnumGene&lt;&gt;(</span>
					m.allele.index,
<span class="fc" id="L295">					ISeq.of(m.alleles)</span>
<span class="fc" id="L296">						.map(jaxb.Unmarshaller(m.allele.value))</span>
				);
			}

		}
	}

}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.1.201405082137</span></div></body></html>