<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../.resources/report.css" type="text/css"/><link rel="shortcut icon" href="../.resources/report.gif" type="image/gif"/><title>Genotype.java</title><link rel="stylesheet" href="../.resources/prettify.css" type="text/css"/><script type="text/javascript" src="../.resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../.sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">org.jenetics</a> &gt; <a href="index.source.html" class="el_package">org.jenetics</a> &gt; <span class="el_source">Genotype.java</span></div><h1>Genotype.java</h1><pre class="source lang-java linenums">/*
 * Java Genetic Algorithm Library (@__identifier__@).
 * Copyright (c) @__year__@ Franz Wilhelmstötter
 *
 * Licensed under the Apache License, Version 2.0 (the &quot;License&quot;);
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an &quot;AS IS&quot; BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Author:
 *    Franz Wilhelmstötter (franz.wilhelmstoetter@gmx.at)
 */
package org.jenetics;

import static org.jenetics.internal.util.Equality.eq;
import static org.jenetics.util.ISeq.toISeq;

import java.io.Serializable;
import java.util.Iterator;
import java.util.List;
import java.util.stream.Stream;

import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorType;
import javax.xml.bind.annotation.XmlAttribute;
import javax.xml.bind.annotation.XmlElement;
import javax.xml.bind.annotation.XmlRootElement;
import javax.xml.bind.annotation.XmlType;
import javax.xml.bind.annotation.adapters.XmlAdapter;
import javax.xml.bind.annotation.adapters.XmlJavaTypeAdapter;

import org.jenetics.internal.util.Hash;
import org.jenetics.internal.util.jaxb;
import org.jenetics.internal.util.reflect;

import org.jenetics.util.Factory;
import org.jenetics.util.ISeq;
import org.jenetics.util.MSeq;
import org.jenetics.util.Seq;
import org.jenetics.util.Verifiable;

/**
 * The central class the GA is working with, is the {@code Genotype}. It is the
 * structural representative of an individual. This class is the encoded problem
 * solution with one to many {@link Chromosome}.
 * &lt;p&gt;
 * &lt;img alt=&quot;Genotype&quot; src=&quot;doc-files/Genotype.svg&quot; width=&quot;400&quot; height=&quot;252&quot; &gt;
 * &lt;/p&gt;
 * The chromosomes of a genotype doesn't have to have necessarily the same size.
 * It is only required that all genes are from the same type and the genes within
 * a chromosome have the same constraints; e. g. the same min- and max values
 * for number genes.
 *
 * &lt;pre&gt;{@code
 * final Genotype&lt;DoubleGene&gt; genotype = Genotype.of(
 *     DoubleChromosome.of(0.0, 1.0, 8),
 *     DoubleChromosome.of(1.0, 2.0, 10),
 *     DoubleChromosome.of(0.0, 10.0, 9),
 *     DoubleChromosome.of(0.1, 0.9, 5)
 * );
 * }&lt;/pre&gt;
 * The code snippet above creates a genotype with the same structure as shown in
 * the figure above. In this example the {@link DoubleGene} has been chosen as
 * gene type.
 *
 * @author &lt;a href=&quot;mailto:franz.wilhelmstoetter@gmx.at&quot;&gt;Franz Wilhelmstötter&lt;/a&gt;
 * @since 1.0
 * @version 3.0
 */
<span class="pc bpc" id="L77" title="1 of 2 branches missed.">@XmlJavaTypeAdapter(Genotype.Model.Adapter.class)</span>
public final class Genotype&lt;G extends Gene&lt;?, G&gt;&gt;
	implements
		Factory&lt;Genotype&lt;G&gt;&gt;,
		Iterable&lt;Chromosome&lt;G&gt;&gt;,
		Verifiable,
		Serializable
{
	private static final long serialVersionUID = 3L;

	private final ISeq&lt;Chromosome&lt;G&gt;&gt; _chromosomes;
	private final int _ngenes;

	//Caching isValid value.
<span class="fc" id="L91">	private volatile Boolean _valid = null;</span>

	private Genotype(
		final ISeq&lt;? extends Chromosome&lt;G&gt;&gt; chromosomes,
		final int ngenes
<span class="fc" id="L96">	) {</span>
<span class="pc bpc" id="L97" title="1 of 2 branches missed.">		if (chromosomes.isEmpty()) {</span>
<span class="nc" id="L98">			throw new IllegalArgumentException(&quot;No chromosomes given.&quot;);</span>
		}

<span class="fc" id="L101">		_chromosomes = reflect.cast(chromosomes);</span>
<span class="fc" id="L102">		_ngenes = ngenes;</span>
<span class="fc" id="L103">	}</span>

	/**
	 * Create a new Genotype from a given sequence of {@code Chromosomes}.
	 *
	 * @param chromosomes The {@code Chromosome} array the {@code Genotype}
	 *         consists of.
	 * @throws NullPointerException if {@code chromosomes} is null or one of its
	 *         element.
	 * @throws IllegalArgumentException if {@code chromosome.length == 0}.
	 */
	Genotype(final ISeq&lt;? extends Chromosome&lt;G&gt;&gt; chromosomes) {
<span class="fc" id="L115">		this(chromosomes, ngenes(chromosomes));</span>
<span class="fc" id="L116">	}</span>

	private static int ngenes(final Seq&lt;? extends Chromosome&lt;?&gt;&gt; chromosomes) {
<span class="fc" id="L119">		return chromosomes.stream()</span>
<span class="fc" id="L120">			.mapToInt(c -&gt; c.length())</span>
<span class="fc" id="L121">			.sum();</span>
	}

	/**
	 * Return the chromosome at the given index. It is guaranteed, that the
	 * returned chromosome is not null.
	 *
	 * @param index Chromosome index.
	 * @return The Chromosome.
	 * @throws IndexOutOfBoundsException if
	 *         {@code (index &lt; 0 || index &gt;= _length)}.
	 */
	public Chromosome&lt;G&gt; getChromosome(final int index) {
<span class="pc bpc" id="L134" title="2 of 4 branches missed.">		assert _chromosomes != null;</span>
<span class="pc bpc" id="L135" title="2 of 4 branches missed.">		assert _chromosomes.get(index) != null;</span>

<span class="fc" id="L137">		return _chromosomes.get(index);</span>
	}

	/**
	 * Return the first chromosome. This is a shortcut for
	 * &lt;pre&gt;{@code
	 * final Genotype&lt;DoubleGene&gt;; gt = ...
	 * final Chromosome&lt;DoubleGene&gt; chromosome = gt.getChromosome(0);
	 * }&lt;/pre&gt;
	 *
	 * @return The first chromosome.
	 */
	public Chromosome&lt;G&gt; getChromosome() {
<span class="pc bpc" id="L150" title="2 of 4 branches missed.">		assert _chromosomes != null;</span>
<span class="pc bpc" id="L151" title="2 of 4 branches missed.">		assert _chromosomes.get(0) != null;</span>

<span class="fc" id="L153">		return _chromosomes.get(0);</span>
	}

	/**
	 * Return the first {@link Gene} of the first {@link Chromosome} of this
	 * {@code Genotype}. This is a shortcut for
	 * &lt;pre&gt;{@code
	 * final Genotype&lt;DoubleGene&gt; gt = ...
	 * final DoubleGene gene = gt.getChromosome(0).getGene(0);
	 * }&lt;/pre&gt;
	 *
	 * @return the first {@link Gene} of the first {@link Chromosome} of this
	 *         {@code Genotype}.
	 */
	public G getGene() {
<span class="pc bpc" id="L168" title="2 of 4 branches missed.">		assert _chromosomes != null;</span>
<span class="pc bpc" id="L169" title="2 of 4 branches missed.">		assert _chromosomes.get(0) != null;</span>

<span class="fc" id="L171">		return _chromosomes.get(0).getGene();</span>
	}

	/**
	 * Return the gene from the given chromosome- and gene index. This is a
	 * shortcut for {@code gt.getChromosome(chromosomeIndex).getGene(geneIndex)}.
	 *
	 * @since 3.0
	 *
	 * @param chromosomeIndex the chromosome index
	 * @param geneIndex the gene index within the chromosome
	 * @return the gene with the given indexes
	 * @throws IndexOutOfBoundsException if the given indexes are not within the
	 *         allowed range
	 */
	public G get(final int chromosomeIndex, final int geneIndex) {
<span class="fc" id="L187">		return getChromosome(chromosomeIndex).getGene(geneIndex);</span>
	}

	public ISeq&lt;Chromosome&lt;G&gt;&gt; toSeq() {
<span class="fc" id="L191">		return _chromosomes;</span>
	}

	@Override
	public Iterator&lt;Chromosome&lt;G&gt;&gt; iterator() {
<span class="nc" id="L196">		return _chromosomes.iterator();</span>
	}

	/**
	 * Returns a sequential {@code Stream} of chromosomes with this genotype as
	 * its source.
	 *
	 * @since !__version__!
	 *
	 * @return a sequential {@code Stream} of chromosomes
	 */
	public Stream&lt;Chromosome&lt;G&gt;&gt; stream() {
<span class="fc" id="L208">		return _chromosomes.stream();</span>
	}

	/**
	 * Getting the number of chromosomes of this genotype.
	 *
	 * @return number of chromosomes.
	 */
	public int length() {
<span class="fc" id="L217">		return _chromosomes.length();</span>
	}

	/**
	 * Return the number of genes this genotype consists of. This is the sum of
	 * the number of genes of the genotype chromosomes.
	 *
	 * @return Return the number of genes this genotype consists of.
	 */
	public int getNumberOfGenes() {
<span class="fc" id="L227">		return _ngenes;</span>
	}

	/**
	 * Test if this genotype is valid. A genotype is valid if all its
	 * {@link Chromosome}s are valid.
	 *
	 * @return true if this genotype is valid, false otherwise.
	 */
	@Override
	public boolean isValid() {
<span class="fc" id="L238">		Boolean valid = _valid;</span>
<span class="fc bfc" id="L239" title="All 2 branches covered.">		if (valid == null) {</span>
<span class="fc" id="L240">			valid = _chromosomes.forAll(Verifiable::isValid);</span>
<span class="fc" id="L241">			_valid = valid;</span>
		}

<span class="fc" id="L244">		return _valid;</span>
	}

	/**
	 * Return a new, random genotype by creating new, random chromosomes (calling
	 * the {@link Chromosome#newInstance()} method) from the chromosomes of this
	 * genotype.
	 */
	@Override
	public Genotype&lt;G&gt; newInstance() {
<span class="fc" id="L254">		return new Genotype&lt;&gt;(_chromosomes.map(Factory::newInstance), _ngenes);</span>
	}

	Genotype&lt;G&gt; newInstance(final ISeq&lt;Chromosome&lt;G&gt;&gt; chromosomes) {
<span class="fc" id="L258">		return new Genotype&lt;&gt;(chromosomes, _ngenes);</span>
	}

	@Override
	public int hashCode() {
<span class="fc" id="L263">		return Hash.of(getClass()).and(_chromosomes).value();</span>
	}

	@Override
	public boolean equals(final Object obj) {
<span class="fc bfc" id="L268" title="All 2 branches covered.">		return obj instanceof Genotype&lt;?&gt; &amp;&amp;</span>
<span class="fc bfc" id="L269" title="All 2 branches covered.">			eq(_chromosomes, ((Genotype&lt;?&gt;)obj)._chromosomes);</span>
	}

	@Override
	public String toString() {
<span class="fc" id="L274">		return _chromosomes.toString();</span>
	}

	/**
	 * Create a new {@code Genotype} from a given array of {@code Chromosomes}.
	 *
	 * @since 3.0
	 *
	 * @param &lt;G&gt; the gene type
	 * @param first the first {@code Chromosome} of the {@code Genotype}
	 * @param rest the rest of the genotypes chromosomes.
	 * @return a new {@code Genotype} from the given chromosomes
	 * @throws NullPointerException if {@code chromosomes} is {@code null} or
	 *         one of its element.
	 */
	@SafeVarargs
	public static &lt;G extends Gene&lt;?, G&gt;&gt; Genotype&lt;G&gt; of(
		final Chromosome&lt;G&gt; first,
		final Chromosome&lt;G&gt;... rest
	) {
<span class="fc" id="L294">		final MSeq&lt;Chromosome&lt;G&gt;&gt; seq = MSeq.ofLength(1 + rest.length);</span>
<span class="fc" id="L295">		seq.set(0, first);</span>
<span class="fc bfc" id="L296" title="All 2 branches covered.">		for (int i = 0; i &lt; rest.length; ++i) {</span>
<span class="fc" id="L297">			seq.set(i + 1, rest[i]);</span>
		}
<span class="fc" id="L299">		return new Genotype&lt;&gt;(seq.toISeq());</span>
	}

	/**
	 * Create a new {@code Genotype} which consists of {@code n} chromosomes,
	 * which are created by the given {@code factory}. This method can be used
	 * for easily creating a &lt;i&gt;gene matrix&lt;/i&gt;. The following example will
	 * create a 10x5 {@code DoubleGene} &lt;i&gt;matrix&lt;/i&gt;.
	 *
	 * &lt;pre&gt;{@code
	 * final Genotype&lt;DoubleGene&gt; gt = Genotype
	 *     .of(DoubleChromosome.of(0.0, 1.0, 10), 5);
	 * }&lt;/pre&gt;
	 *
	 * @since 3.0
	 *
	 * @param &lt;G&gt; the gene type
	 * @param factory the factory which creates the chromosomes this genotype
	 *        consists of
	 * @param n the number of chromosomes this genotype consists of
	 * @return new {@code Genotype} containing {@code n} chromosomes
	 * @throws IllegalArgumentException if {@code n &lt; 1}.
	 * @throws NullPointerException if the {@code factory} is {@code null}.
	 */
	public static &lt;G extends Gene&lt;?, G&gt;&gt; Genotype&lt;G&gt;
	of(final Factory&lt;? extends Chromosome&lt;G&gt;&gt; factory, final int n) {
<span class="nc" id="L325">		final ISeq&lt;Chromosome&lt;G&gt;&gt; ch = ISeq.of(factory::newInstance, n);</span>
<span class="nc" id="L326">		return new Genotype&lt;&gt;(ch);</span>
	}

	/**
	 * Create a new {@code Genotype} from a given array of {@code chromosomes}.
	 *
	 * @since 3.0
	 *
	 * @param &lt;G&gt; the gene type
	 * @param chromosomes the {@code Chromosome}s the returned genotype consists
	 *        of
	 * @return a new {@code Genotype} from the given chromosomes
	 * @throws NullPointerException if {@code chromosomes} is {@code null} or
	 *         one of its element.
	 * @throws IllegalArgumentException if {@code chromosome.length() &lt; 1}.
	 */
	public static &lt;G extends Gene&lt;?, G&gt;&gt; Genotype&lt;G&gt;
	of(final Iterable&lt;? extends Chromosome&lt;G&gt;&gt; chromosomes) {
<span class="fc" id="L344">		return new Genotype&lt;&gt;(ISeq.of(chromosomes));</span>
	}

	/* *************************************************************************
	 *  JAXB object serialization
	 * ************************************************************************/

	@XmlRootElement(name = &quot;genotype&quot;)
	@XmlType(name = &quot;org.jenetics.Genotype&quot;)
	@XmlAccessorType(XmlAccessType.FIELD)
	@SuppressWarnings({&quot;unchecked&quot;, &quot;rawtypes&quot;})
<span class="fc" id="L355">	static final class Model {</span>

		@XmlAttribute(name = &quot;length&quot;, required = true)
		public int length;

		@XmlAttribute(name = &quot;ngenes&quot;, required = true)
		public int ngenes;

		@XmlElement(name = &quot;chromosome&quot;, required = true, nillable = false)
		public List chromosomes;

<span class="fc" id="L366">		public static final class Adapter</span>
			extends XmlAdapter&lt;Model, Genotype&gt;
		{
			@Override
			public Model marshal(final Genotype gt) throws Exception {
<span class="nc" id="L371">				final Model model = new Model();</span>
<span class="nc" id="L372">				model.length = gt.length();</span>
<span class="nc" id="L373">				model.ngenes = gt.getNumberOfGenes();</span>
<span class="nc" id="L374">				model.chromosomes = gt.toSeq()</span>
<span class="nc" id="L375">					.map(jaxb.Marshaller(gt.getChromosome()))</span>
<span class="nc" id="L376">					.asList();</span>

<span class="nc" id="L378">				return model;</span>
			}

			@Override
			public Genotype unmarshal(final Model model) throws Exception {
<span class="fc" id="L383">				final ISeq chs = (ISeq)model.chromosomes.stream()</span>
<span class="fc" id="L384">					.map(jaxb.Unmarshaller(model.chromosomes.get(0)))</span>
<span class="fc" id="L385">					.collect(toISeq());</span>

<span class="fc" id="L387">				return new Genotype(chs, model.ngenes);</span>
			}
		}

<span class="fc" id="L391">		public static final Adapter ADAPTER = new Adapter();</span>
	}
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.1.201405082137</span></div></body></html>