<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../.resources/report.css" type="text/css"/><link rel="shortcut icon" href="../.resources/report.gif" type="image/gif"/><title>LongChromosome.java</title><link rel="stylesheet" href="../.resources/prettify.css" type="text/css"/><script type="text/javascript" src="../.resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../.sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">org.jenetics</a> &gt; <a href="index.source.html" class="el_package">org.jenetics</a> &gt; <span class="el_source">LongChromosome.java</span></div><h1>LongChromosome.java</h1><pre class="source lang-java linenums">/*
 * Java Genetic Algorithm Library (@__identifier__@).
 * Copyright (c) @__year__@ Franz Wilhelmstötter
 *
 * Licensed under the Apache License, Version 2.0 (the &quot;License&quot;);
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an &quot;AS IS&quot; BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Author:
 *    Franz Wilhelmstötter (franz.wilhelmstoetter@gmx.at)
 */
package org.jenetics;

import static org.jenetics.util.ISeq.toISeq;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.util.List;

import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorType;
import javax.xml.bind.annotation.XmlAttribute;
import javax.xml.bind.annotation.XmlElement;
import javax.xml.bind.annotation.XmlRootElement;
import javax.xml.bind.annotation.XmlType;
import javax.xml.bind.annotation.adapters.XmlAdapter;
import javax.xml.bind.annotation.adapters.XmlJavaTypeAdapter;

import org.jenetics.internal.util.Equality;
import org.jenetics.internal.util.Hash;

import org.jenetics.util.ISeq;
import org.jenetics.util.LongRange;
import org.jenetics.util.MSeq;

/**
 * Numeric chromosome implementation which holds 64 bit integer numbers.
 *
 * @author &lt;a href=&quot;mailto:franz.wilhelmstoetter@gmx.at&quot;&gt;Franz Wilhelmstötter&lt;/a&gt;
 * @since 1.6
 * @version 3.2
 */
@XmlJavaTypeAdapter(LongChromosome.Model.Adapter.class)
public class LongChromosome
	extends AbstractBoundedChromosome&lt;Long, LongGene&gt;
	implements
		NumericChromosome&lt;Long, LongGene&gt;,
		Serializable
{
	private static final long serialVersionUID = 1L;


	protected LongChromosome(final ISeq&lt;LongGene&gt; genes) {
<span class="fc" id="L64">		super(genes);</span>
<span class="fc" id="L65">	}</span>

	/**
	 * Create a new random {@code LongChromosome}.
	 *
	 * @param min the min value of the {@link LongGene}s (inclusively).
	 * @param max the max value of the {@link LongGene}s (inclusively).
	 * @param length the length of the chromosome.
	 * @throws NullPointerException if one of the arguments is {@code null}.
	 */
	public LongChromosome(final Long min, final Long max, final int length) {
<span class="fc" id="L76">		this(LongGene.seq(min, max, length));</span>
<span class="fc" id="L77">		_valid = true;</span>
<span class="fc" id="L78">	}</span>

	/**
	 * Create a new random {@code LongChromosome} of length one.
	 *
	 * @param min the minimal value of this chromosome (inclusively).
	 * @param max the maximal value of this chromosome (inclusively).
	 * @throws NullPointerException if one of the arguments is {@code null}.
	 */
	public LongChromosome(final Long min, final Long max) {
<span class="fc" id="L88">		this(min, max, 1);</span>
<span class="fc" id="L89">	}</span>

	/**
	 * Returns an long array containing all of the elements in this chromosome
	 * in proper sequence.  If the chromosome fits in the specified array, it is
	 * returned therein. Otherwise, a new array is allocated with the length of
	 * this chromosome.
	 *
	 * @since 3.0
	 *
	 * @param array the array into which the elements of this chromosomes are to
	 *        be stored, if it is big enough; otherwise, a new array is
	 *        allocated for this purpose.
	 * @return an array containing the elements of this chromosome
	 * @throws NullPointerException if the given {@code array} is {@code null}
	 */
	public long[] toArray(final long[] array) {
<span class="pc bpc" id="L106" title="1 of 2 branches missed.">		final long[] a = array.length &gt;= length() ?</span>
<span class="pc" id="L107">			array : new long[length()];</span>

<span class="fc bfc" id="L109" title="All 2 branches covered.">		for (int i = length(); --i &gt;= 0;) {</span>
<span class="fc" id="L110">			a[i] = longValue(i);</span>
		}

<span class="fc" id="L113">		return a;</span>
	}

	/**
	 * Returns an long array containing all of the elements in this chromosome
	 * in proper sequence.
	 *
	 * @since 3.0
	 *
	 * @return an array containing the elements of this chromosome
	 */
	public long[] toArray() {
<span class="fc" id="L125">		return toArray(new long[length()]);</span>
	}

	/**
	 * Create a new {@code LongChromosome} with the given genes.
	 *
	 * @param genes the genes of the chromosome.
	 * @return a new chromosome with the given genes.
	 * @throws IllegalArgumentException if the length of the genes array is
	 *         empty.
	 */
	public static LongChromosome of(final LongGene... genes) {
<span class="fc" id="L137">		return new LongChromosome(ISeq.of(genes));</span>
	}

	/**
	 * Create a new random {@code LongChromosome}.
	 *
	 * @param min the min value of the {@link LongGene}s (inclusively).
	 * @param max the max value of the {@link LongGene}s (inclusively).
	 * @param length the length of the chromosome.
	 * @return a new {@code LongChromosome} with the given gene parameters.
	 */
	public static LongChromosome of(
		final long min,
		final long max,
		final int length
	) {
<span class="fc" id="L153">		return new LongChromosome(min, max, length);</span>
	}

	/**
	 * Create a new random {@code LongChromosome}.
	 *
	 * @since 3.2
	 *
	 * @param range the long range of the chromosome.
	 * @param length the length of the chromosome.
	 * @return a new random {@code LongChromosome}
	 * @throws NullPointerException if the given {@code range} is {@code null}
	 */
	public static LongChromosome of(final LongRange range, final int length) {
<span class="fc" id="L167">		return new LongChromosome(range.getMin(), range.getMax(), length);</span>
	}

	/**
	 * Create a new random {@code LongChromosome} of length one.
	 *
	 * @param min the minimal value of this chromosome (inclusively).
	 * @param max the maximal value of this chromosome (inclusively).
	 * @return a new {@code LongChromosome} with the given gene parameters.
	 */
	public static LongChromosome of(final long min, final long max) {
<span class="fc" id="L178">		return new LongChromosome(min, max);</span>
	}

	/**
	 * Create a new random {@code LongChromosome} of length one.
	 *
	 * @since 3.2
	 *
	 * @param range the long range of the chromosome.
	 * @return a new random {@code LongChromosome} of length one
	 * @throws NullPointerException if the given {@code range} is {@code null}
	 */
	public static LongChromosome of(final LongRange range) {
<span class="fc" id="L191">		return new LongChromosome(range.getMin(), range.getMax());</span>
	}

	@Override
	public LongChromosome newInstance(final ISeq&lt;LongGene&gt; genes) {
<span class="fc" id="L196">		return new LongChromosome(genes);</span>
	}

	@Override
	public LongChromosome newInstance() {
<span class="fc" id="L201">		return new LongChromosome(_min, _max, length());</span>
	}

	@Override
	public int hashCode() {
<span class="fc" id="L206">		return Hash.of(getClass()).and(super.hashCode()).value();</span>
	}

	@Override
	public boolean equals(final Object obj) {
<span class="fc" id="L211">		return Equality.of(this, obj).test(super::equals);</span>
	}

	/* *************************************************************************
	 *  Java object serialization
	 * ************************************************************************/

	private void writeObject(final ObjectOutputStream out)
		throws IOException
	{
<span class="fc" id="L221">		out.defaultWriteObject();</span>

<span class="fc" id="L223">		out.writeInt(length());</span>
<span class="fc" id="L224">		out.writeLong(_min.longValue());</span>
<span class="fc" id="L225">		out.writeLong(_max.longValue());</span>

<span class="fc bfc" id="L227" title="All 2 branches covered.">		for (LongGene gene : _genes) {</span>
<span class="fc" id="L228">			out.writeLong(gene.getAllele().longValue());</span>
<span class="fc" id="L229">		}</span>
<span class="fc" id="L230">	}</span>

	private void readObject(final ObjectInputStream in)
		throws IOException, ClassNotFoundException
	{
<span class="fc" id="L235">		in.defaultReadObject();</span>

<span class="fc" id="L237">		final MSeq&lt;LongGene&gt; genes = MSeq.ofLength(in.readInt());</span>
<span class="fc" id="L238">		_min = in.readLong();</span>
<span class="fc" id="L239">		_max = in.readLong();</span>

<span class="fc bfc" id="L241" title="All 2 branches covered.">		for (int i = 0; i &lt; genes.length(); ++i) {</span>
<span class="fc" id="L242">			genes.set(i, new LongGene(in.readLong(), _min, _max));</span>
		}

<span class="fc" id="L245">		_genes = genes.toISeq();</span>
<span class="fc" id="L246">	}</span>

	/* *************************************************************************
	 *  JAXB object serialization
	 * ************************************************************************/

	@XmlRootElement(name = &quot;long-chromosome&quot;)
	@XmlType(name = &quot;org.jenetics.LongChromosome&quot;)
	@XmlAccessorType(XmlAccessType.FIELD)
<span class="fc" id="L255">	final static class Model {</span>

		@XmlAttribute(name = &quot;length&quot;, required = true)
		public int length;

		@XmlAttribute(name = &quot;min&quot;, required = true)
		public long min;

		@XmlAttribute(name = &quot;max&quot;, required = true)
		public long max;

		@XmlElement(name = &quot;allele&quot;, required = true, nillable = false)
		public List&lt;Long&gt; values;

<span class="fc" id="L269">		public final static class Adapter</span>
			extends XmlAdapter&lt;Model, LongChromosome&gt;
		{
			@Override
			public Model marshal(final LongChromosome c) {
<span class="nc" id="L274">				final Model m = new Model();</span>
<span class="nc" id="L275">				m.length = c.length();</span>
<span class="nc" id="L276">				m.min = c._min;</span>
<span class="nc" id="L277">				m.max = c._max;</span>
<span class="nc" id="L278">				m.values = c.toSeq().map(LongGene::getAllele).asList();</span>
<span class="nc" id="L279">				return m;</span>
			}

			@Override
			public LongChromosome unmarshal(final Model model) {
<span class="fc" id="L284">				final Long min = model.min;</span>
<span class="fc" id="L285">				final Long max = model.max;</span>
<span class="fc" id="L286">				return new LongChromosome(</span>
<span class="fc" id="L287">					model.values.stream()</span>
<span class="fc" id="L288">						.map(value -&gt; new LongGene(value, min, max))</span>
<span class="fc" id="L289">						.collect(toISeq())</span>
				);
			}
		}
	}
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.1.201405082137</span></div></body></html>